File: DID.Native.SDK/src/jwt/jwtparser.c
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <cjose/cjose.h>
#include <jansson.h>

#include "ela_did.h"
#include "ela_jwt.h"
#include "crypto.h"
#include "HDkey.h"
#include "jws.h"
#include "diderror.h"

static cjose_jwk_t *get_jwk(JWS *jws)
{
    cjose_err err;
    DID *issuer = NULL;
    DIDDocument *doc = NULL;
    DIDURL *keyid;
    PublicKey *key;
    const char *keybase58;
    uint8_t binkey[PUBLICKEY_BYTES];
    KeySpec _spec, *spec;
    cjose_jwk_t *jwk = NULL;
    int rc = -1;

    assert(jws);
    assert(jws->header);
    assert(jws->claims);

    issuer = DID_FromString(JWS_GetIssuer(jws));
    if (!issuer)
        goto errorExit;

    doc = DID_Resolve(issuer, false);
    if (!doc)
        goto errorExit;

    if (!JWS_GetKeyId(jws)) {
        keyid = DIDDocument_GetDefaultPublicKey(doc);
        key = DIDDocument_GetPublicKey(doc, keyid);
    }
    else {
        keyid = DIDURL_FromString(JWS_GetKeyId(jws), issuer);
        key = DIDDocument_GetPublicKey(doc, keyid);
        DIDURL_Destroy(keyid);
    }
    if (!key)
        goto errorExit;

    keybase58 = PublicKey_GetPublicKeyBase58(key);
    if (!keybase58)
        goto errorExit;

    base58_decode(binkey, sizeof(binkey), keybase58);

    memset(&_spec, 0, sizeof(KeySpec));
    spec = KeySpec_Fill(&_spec, binkey, NULL);
    if (!spec) {
        DIDError_Set(DIDERR_CRYPTO_ERROR, "Get key spec failed.");
        goto errorExit;
    }

    jwk = cjose_jwk_create_EC_spec((cjose_jwk_ec_keyspec*)spec, &err);
    if (!jwk) {
        DIDError_Set(DIDERR_JWT, "Create jwk failed.");
        goto errorExit;
    }

errorExit:
    if (issuer)
        DID_Destroy(issuer);
    if (doc)
        DIDDocument_Destroy(doc);

    return jwk;
}

static JWS *parse_jwt(const char *token, int dot)
{
    JWS *jws = NULL;
    char *claims, *header, *_token;
    size_t len;

    assert(token && *token);
    assert(dot > 0 && dot < strlen(token));

    jws = (JWS *)calloc(1, sizeof(JWS));
    if (!jws) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Remalloc buffer for JWS failed.");
        return NULL;
    }

    //copy token
    len = strlen(token);
    _token = (char*)alloca(len);
    strncpy(_token, token, len - 1);
    _token[len - 1] = 0;

    //get claims
    len = strlen(_token) - dot - 1;
    claims = (char*)alloca(len + 1);
    len = base64_url_decode((uint8_t *)claims, _token + dot + 1);
    if (len <= 0) {
        DIDError_Set(DIDERR_CRYPTO_ERROR, "Decode jwt claims failed");
        goto errorExit;
    }
    claims[len] = 0;

    jws->claims = json_loadb(claims, len, 0, NULL);
    if (!jws->claims) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Load jwt body failed.");
        goto errorExit;
    }

    //get header
    _token[dot] = 0;
    len = dot;
    header = (char*)alloca(len + 1);
    len = base64_url_decode((uint8_t *)header, _token);
    if (len <= 0) {
        DIDError_Set(DIDERR_CRYPTO_ERROR, "Decode jwt header failed");
        goto errorExit;
    }

    jws->header = json_loadb(header, len, 0, NULL);
    if (!jws->header) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Load jwt header failed.");
        goto errorExit;
    }

    return jws;

errorExit:
    if (jws)
        JWS_Destroy(jws);

    return NULL;
}

static JWS *parse_jws(const char *token)
{
    JWS *jws = NULL;
    cjose_err err;
    cjose_jwk_t *jwk = NULL;
    cjose_jws_t *jws_t = NULL;
    char *payload = NULL;
    size_t payload_len = 0;
    bool successed;
    DIDURL *signkey;

    assert(token && *token);

    jws = (JWS *)calloc(1, sizeof(JWS));
    if (!jws) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Remalloc buffer for JWS failed.");
        return NULL;
    }

    //set jws
    jws_t = cjose_jws_import(token, strlen(token), &err);
    if (!jws_t) {
        DIDError_Set(DIDERR_JWT, "Import token to jws failed.");
        goto errorExit;
    }

    //get header
    json_t *json = cjose_jws_get_protected(jws_t);
    jws->header = json_deep_copy(cjose_jws_get_protected(jws_t));
    if (!jws->header) {
        DIDError_Set(DIDERR_JWT, "Get jwt header failed.");
        goto errorExit;
    }

    //set claims(payload)
    successed = cjose_jws_get_plaintext(jws_t, (uint8_t**)&payload, &payload_len, &err);
    if (!successed) {
        DIDError_Set(DIDERR_JWT, "Get jwt body failed.");
        goto errorExit;
    }

    jws->claims = json_loadb(payload, payload_len, 0, NULL);;
    if (!jws->claims) {
        DIDError_Set(DIDERR_OUT_OF_MEMORY, "Load jwt body failed.");
        goto errorExit;
    }

    //get jwk, must put after getting header and claims.
    jwk = get_jwk(jws);
    if (!jwk) {
        JWS_Destroy(jws);
        return NULL;
    }

    successed = cjose_jws_verify(jws_t, jwk, &err);
    cjose_jwk_release(jwk);
    if (!successed) {
        DIDError_Set(DIDERR_JWT, "Verify jws failed.");
        JWS_Destroy(jws);
        return NULL;
    }
    return jws;

errorExit:
    if (jws_t)
        cjose_jws_release(jws_t);
    if (jws)
        JWS_Destroy(jws);

    return NULL;
}

JWS *JWTParser_Parse(const char *token)
{
    int idx = 0;
    int dots[2] = {0, 0};

    if (!token || !*token) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Invalid arguments.");
        return NULL;
    }

    // find the indexes of the dots
    for (int i = 0; i < strlen(token) && idx < 2; ++i) {
        if (token[i] == '.')
            dots[idx++] = i;
    }

    if (idx != 2 || !(dots[0] > 0 && dots[1] < strlen(token))) {
        DIDError_Set(DIDERR_INVALID_ARGS, "Invalid jwt token! Please check it.");
        return NULL;
    }

    if (dots[1] == strlen(token) - 1)
        return parse_jwt(token, dots[0]);

    return parse_jws(token);
}

File: DID.Java.SDK/src/main/java/org/elastos/did/jwt/JwtParser.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did.jwt;

public class JwtParser {
	private io.jsonwebtoken.JwtParser impl;

	protected JwtParser(io.jsonwebtoken.JwtParser impl) {
		this.impl = impl;
	}

	/**
	 * Returns {@code true} if the specified JWT compact string represents a
	 * signed JWT (aka a 'JWS'), {@code false} otherwise.
	 * <p>
	 * <p>
	 * Note that if you are reasonably sure that the token is signed, it is more
	 * efficient to attempt to parse the token (and catching exceptions if
	 * necessary) instead of calling this method first before parsing.
	 * </p>
	 *
	 * @param jwt the compact serialized JWT to check
	 * @return {@code true} if the specified JWT compact string represents a
	 *         signed JWT (aka a 'JWS'), {@code false} otherwise.
	 */
	public boolean isSigned(String jwt) {
		return impl.isSigned(jwt);
	}

	/**
	 * Parses the specified compact serialized JWT string based on the builder's
	 * current configuration state and returns the resulting JWT or JWS
	 * instance.
	 * <p>
	 * <p>
	 * This method returns a JWT or JWS based on the parsed string.
	 *
	 * @param jwt the compact serialized JWT to parse
	 * @return the specified compact serialized JWT string based on the
	 *         builder's current configuration state.
	 * @throws MalformedJwtException    if the specified JWT was incorrectly
	 *                                  constructed (and therefore invalid).
	 *                                  Invalid JWTs should not be trusted and
	 *                                  should be discarded.
	 * @throws JwsSignatureException    if a JWS signature was discovered, but
	 *                                  could not be verified. JWTs that fail
	 *                                  signature validation should not be
	 *                                  trusted and should be discarded.
	 * @throws ExpiredJwtException      if the specified JWT is a Claims JWT and
	 *                                  the Claims has an expiration time before
	 *                                  the time this method is invoked.
	 * @throws IllegalArgumentException if the specified string is {@code null}
	 *                                  or empty or only whitespace.
	 * @see #parsePlaintextJwt(String)
	 * @see #parseClaimsJwt(String)
	 * @see #parsePlaintextJws(String)
	 * @see #parseClaimsJws(String)
	 */
	public Jwt<?> parse(String jwt) throws ExpiredJwtException,
			MalformedJwtException, JwsSignatureException {
		try {
			return new Jwt<String>(impl.parse(jwt));
		} catch (io.jsonwebtoken.ExpiredJwtException e) {
			throw new ExpiredJwtException(e);
		} catch (io.jsonwebtoken.MalformedJwtException e) {
			throw new MalformedJwtException(e);
		} catch (@SuppressWarnings("deprecation") io.jsonwebtoken.SignatureException e) {
			throw new JwsSignatureException(e);
		}
	}

	/**
	 * Parses the specified compact serialized JWT string based on the builder's
	 * current configuration state and returns the resulting unsigned plaintext
	 * JWT instance.
	 * <p>
	 * <p>
	 * This is a convenience method that is usable if you are confident that the
	 * compact string argument reflects an unsigned plaintext JWT. An unsigned
	 * plaintext JWT has a String (non-JSON) body payload and it is not
	 * cryptographically signed.
	 * </p>
	 * <p>
	 * <p>
	 * <b>If the compact string presented does not reflect an unsigned plaintext
	 * JWT with non-JSON string body, an {@link UnsupportedJwtException} will be
	 * thrown.</b>
	 * </p>
	 *
	 * @param plaintextJwt a compact serialized unsigned plaintext JWT string.
	 * @return the {@link Jwt Jwt} instance that reflects the specified compact
	 *         JWT string.
	 * @throws UnsupportedJwtException  if the {@code plaintextJwt} argument
	 *                                  does not represent an unsigned plaintext
	 *                                  JWT
	 * @throws MalformedJwtException    if the {@code plaintextJwt} string is
	 *                                  not a valid JWT
	 * @throws JwsSignatureException    if the {@code plaintextJwt} string is
	 *                                  actually a JWS and signature validation
	 *                                  fails
	 * @throws IllegalArgumentException if the {@code plaintextJwt} string is
	 *                                  {@code null} or empty or only whitespace
	 * @see #parseClaimsJwt(String)
	 * @see #parsePlaintextJws(String)
	 * @see #parseClaimsJws(String)
	 * @see #parse(String)
	 * @since 0.2
	 */
	public Jwt<String> parsePlaintextJwt(String plaintextJwt)
			throws UnsupportedJwtException, MalformedJwtException,
			JwsSignatureException {
		try {
			return new Jwt<String>(impl.parsePlaintextJwt(plaintextJwt));
		} catch (io.jsonwebtoken.UnsupportedJwtException e) {
			throw new UnsupportedJwtException(e);
		} catch (io.jsonwebtoken.MalformedJwtException e) {
			throw new MalformedJwtException(e);
		} catch (@SuppressWarnings("deprecation") io.jsonwebtoken.SignatureException e) {
			throw new JwsSignatureException(e);
		}
	}

	/**
	 * Parses the specified compact serialized JWT string based on the builder's
	 * current configuration state and returns the resulting unsigned plaintext
	 * JWT instance.
	 * <p>
	 * <p>
	 * This is a convenience method that is usable if you are confident that the
	 * compact string argument reflects an unsigned Claims JWT. An unsigned
	 * Claims JWT has a {@link Claims} body and it is not cryptographically
	 * signed.
	 * </p>
	 * <p>
	 * <p>
	 * <b>If the compact string presented does not reflect an unsigned Claims
	 * JWT, an {@link UnsupportedJwtException} will be thrown.</b>
	 * </p>
	 *
	 * @param claimsJwt a compact serialized unsigned Claims JWT string.
	 * @return the {@link Jwt Jwt} instance that reflects the specified compact
	 *         JWT string.
	 * @throws UnsupportedJwtException  if the {@code claimsJwt} argument does
	 *                                  not represent an unsigned Claims JWT
	 * @throws MalformedJwtException    if the {@code claimsJwt} string is not a
	 *                                  valid JWT
	 * @throws JwsSignatureException    if the {@code claimsJwt} string is
	 *                                  actually a JWS and signature validation
	 *                                  fails
	 * @throws ExpiredJwtException      if the specified JWT is a Claims JWT and
	 *                                  the Claims has an expiration time before
	 *                                  the time this method is invoked.
	 * @throws IllegalArgumentException if the {@code claimsJwt} string is
	 *                                  {@code null} or empty or only whitespace
	 * @see #parsePlaintextJwt(String)
	 * @see #parsePlaintextJws(String)
	 * @see #parseClaimsJws(String)
	 * @see #parse(String)
	 * @since 0.2
	 */
	public Jwt<Claims> parseClaimsJwt(String claimsJwt)
			throws ExpiredJwtException, UnsupportedJwtException,
			MalformedJwtException, JwsSignatureException {
		try {
			return new Jwt<Claims>(impl.parseClaimsJwt(claimsJwt));
		} catch (io.jsonwebtoken.ExpiredJwtException e) {
			throw new ExpiredJwtException(e);
		} catch (io.jsonwebtoken.UnsupportedJwtException e) {
			throw new UnsupportedJwtException(e);
		} catch (io.jsonwebtoken.MalformedJwtException e) {
			throw new MalformedJwtException(e);
		} catch (@SuppressWarnings("deprecation") io.jsonwebtoken.SignatureException e) {
			throw new JwsSignatureException(e);
		}
	}

	/**
	 * Parses the specified compact serialized JWS string based on the builder's
	 * current configuration state and returns the resulting plaintext JWS
	 * instance.
	 * <p>
	 * <p>
	 * This is a convenience method that is usable if you are confident that the
	 * compact string argument reflects a plaintext JWS. A plaintext JWS is a
	 * JWT with a String (non-JSON) body (payload) that has been
	 * cryptographically signed.
	 * </p>
	 * <p>
	 * <p>
	 * <b>If the compact string presented does not reflect a plaintext JWS, an
	 * {@link UnsupportedJwtException} will be thrown.</b>
	 * </p>
	 *
	 * @param plaintextJws a compact serialized JWS string.
	 * @return the {@link Jws Jws} instance that reflects the specified compact
	 *         JWS string.
	 * @throws UnsupportedJwtException  if the {@code plaintextJws} argument
	 *                                  does not represent an plaintext JWS
	 * @throws MalformedJwtException    if the {@code plaintextJws} string is
	 *                                  not a valid JWS
	 * @throws JwsSignatureException    if the {@code plaintextJws} JWS
	 *                                  signature validation fails
	 * @throws IllegalArgumentException if the {@code plaintextJws} string is
	 *                                  {@code null} or empty or only whitespace
	 * @see #parsePlaintextJwt(String)
	 * @see #parseClaimsJwt(String)
	 * @see #parseClaimsJws(String)
	 * @see #parse(String)
	 * @since 0.2
	 */
	public Jws<String> parsePlaintextJws(String plaintextJws)
			throws UnsupportedJwtException, MalformedJwtException,
			JwsSignatureException {
		try {
			return new Jws<String>(impl.parsePlaintextJws(plaintextJws));
		} catch (io.jsonwebtoken.UnsupportedJwtException e) {
			throw new UnsupportedJwtException(e);
		} catch (io.jsonwebtoken.MalformedJwtException e) {
			throw new MalformedJwtException(e);
		} catch (@SuppressWarnings("deprecation") io.jsonwebtoken.SignatureException e) {
			throw new JwsSignatureException(e);
		}
	}

	/**
	 * Parses the specified compact serialized JWS string based on the builder's
	 * current configuration state and returns the resulting Claims JWS
	 * instance.
	 * <p>
	 * <p>
	 * This is a convenience method that is usable if you are confident that the
	 * compact string argument reflects a Claims JWS. A Claims JWS is a JWT with
	 * a {@link Claims} body that has been cryptographically signed.
	 * </p>
	 * <p>
	 * <p>
	 * <b>If the compact string presented does not reflect a Claims JWS, an
	 * {@link UnsupportedJwtException} will be thrown.</b>
	 * </p>
	 *
	 * @param claimsJws a compact serialized Claims JWS string.
	 * @return the {@link Jws Jws} instance that reflects the specified compact
	 *         Claims JWS string.
	 * @throws UnsupportedJwtException  if the {@code claimsJws} argument does
	 *                                  not represent an Claims JWS
	 * @throws MalformedJwtException    if the {@code claimsJws} string is not a
	 *                                  valid JWS
	 * @throws JwsSignatureException    if the {@code claimsJws} JWS signature
	 *                                  validation fails
	 * @throws ExpiredJwtException      if the specified JWT is a Claims JWT and
	 *                                  the Claims has an expiration time before
	 *                                  the time this method is invoked.
	 * @throws IllegalArgumentException if the {@code claimsJws} string is
	 *                                  {@code null} or empty or only whitespace
	 * @see #parsePlaintextJwt(String)
	 * @see #parseClaimsJwt(String)
	 * @see #parsePlaintextJws(String)
	 * @see #parse(String)
	 * @since 0.2
	 */
	public Jws<Claims> parseClaimsJws(String claimsJws)
			throws ExpiredJwtException, UnsupportedJwtException,
			MalformedJwtException, JwsSignatureException {
		try {
			return new Jws<Claims>(impl.parseClaimsJws(claimsJws));
		} catch (io.jsonwebtoken.ExpiredJwtException e) {
			throw new ExpiredJwtException(e);
		} catch (io.jsonwebtoken.UnsupportedJwtException e) {
			throw new UnsupportedJwtException(e);
		} catch (io.jsonwebtoken.MalformedJwtException e) {
			throw new MalformedJwtException(e);
		} catch (@SuppressWarnings("deprecation") io.jsonwebtoken.SignatureException e) {
			throw new JwsSignatureException(e);
		}
	}
}
